Add support for cross-compiling Swift stdlib with an external sysroot

xtremekforever opened the pull request:
This fixes #75344 and goes towards #78960.

So, here's what I have for building just the Swift stdlib for armv7. I will note that this also works for cross-compiling the same to aarch64 and x86_64, as long as you have the right sysroot.

```bash
./swift/utils/build-script --cross-compile-hosts=linux-armv7 --cross-compile-sysroots=sysroot-debian-bookworm \
     --build-llvm=0  --skip-local-build --build-swift-tools=0 --swift-enable-backtracing=0 --build-embedded-stdlib=0 \
     --native-swift-tools-path=$SWIFT_NATIVE_PATH --native-clang-tools-path=$SWIFT_NATIVE_PATH 
```

With the new `--cross-compile-sysroots` param, we now enable the following functionalities:

 - llvm and cmark will now try to configure themselves with the passed sysroot along with the passed target.
 - The swift stdlib will use the sysroot to compile and pass on flags to the libdispatch build that is used to build Swift Concurrency.

MaxDesiatov committed:
Merge pull request #78909 from xtremekforever/#75341-fix-swift-build-support-typo

Fix typo in get_linux_sysroot that makes cross-compilation fail

xtremekforever committed:
Set system processor in Libdispatch.cmake for LINUX sdks

 - This enables libdispatch to be built for the arch passed and not neccessarily the host arch.

xtremekforever committed:
Add ability to compile Swift stdlib with external sysroot

 - The name CROSS_COMPILE_SYSROOT is chosen since any time we use a sysroot not at "/" we are cross-compiling for some platform or arch.

xtremekforever committed:
Add cross-compile-sysroot option to build scripts

xtremekforever committed:
Only include CROSS_COMPILE_SYSROOT if cross-compiling

xtremekforever committed:
Use lld in product.py if cross_compile_sysroot is used

xtremekforever committed:
Use cross_compile_sysroot in get_linux_sysroot if set

 - This forces it to use the external sysroot instead of the default

xtremekforever committed:
Fix linting issues

xtremekforever commented:
@MaxDesiatov here is my first PR that adds support for building the Swift stdlib using an external sysroot, using a new flag called `--cross-compile-sysroot`.

xtremekforever commented:
To build this inside of a docker container (with Jammy as an example), assuming you have the Swift workspace checked out with my branch:

```bash
docker run -it --name swift-armv7-builder -v $(pwd):/src swiftlang/swift:nightly-jammy /bin/bash
apt update && apt install cmake ninja-build libsqlite3-dev ncurses-dev git python3 libstdc++-11-dev-armhf-cross
cd /src
export SWIFT_NATIVE_PATH=/usr/bin
./swift/utils/build-script --cross-compile-hosts=linux-armv7 --cross-compile-sysroot=/ \
     --build-llvm=0  --skip-local-build --build-swift-tools=0 --swift-enable-backtracing=0 --build-embedded-stdlib=0 \
     --native-swift-tools-path=$SWIFT_NATIVE_PATH --native-clang-tools-path=$SWIFT_NATIVE_PATH 
```

This will cross-compile for armv7 using the cross libraries installed with `libstdc++-11-dev-armhf-cross`. This is an easy way to try it out and works under macOS and Linux alike.

xtremekforever committed:
Find gcc-install-dir in CROSS_COMPILE_SYSROOT for better locating of libstdc++

xtremekforever committed:
Open up cross-compilation to all linux- hosts

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

xtremekforever committed:
Fix tests for llvm_linux_cross_compile support, add more tests

xtremekforever committed:
Merge branch 'main' into #78960-swift-stdlib-cross-compile

xtremekforever commented:
I fixed the tests and added some more as well. I had to merge in main due to conflicts in `product.py` and the `test_llvm_linux_cross_compile.py` that I had to modify. However, now cross-compilation to armv7 and aarch64 fails due to some errors:

```
/src/swift/stdlib/public/core/Slab.swift:108:32: error: invalid conversion from throwing function of type '(Builtin.RawPointer) throws(E) -> ()' to non-throwing function type '(Builtin.RawPointer) -> ()'
106 |   @_alwaysEmitIntoClient
107 |   public init<E: Error>(_ body: (Int) throws(E) -> Element) throws(E) {
108 |     self = try Builtin.emplace { (rawPtr) throws(E) -> () in
    |                                `- error: invalid conversion from throwing function of type '(Builtin.RawPointer) throws(E) -> ()' to non-throwing function type '(Builtin.RawPointer) -> ()'
109 |       let buffer = Slab<count, Element>._initializationBuffer(start: rawPtr)
110 | 

/src/swift/stdlib/public/core/Slab.swift:156:32: error: invalid conversion from throwing function of type '(Builtin.RawPointer) throws(E) -> ()' to non-throwing function type '(Builtin.RawPointer) -> ()'
154 |     var o: Element? = first
155 | 
156 |     self = try Builtin.emplace { (rawPtr) throws(E) -> () in
    |                                `- error: invalid conversion from throwing function of type '(Builtin.RawPointer) throws(E) -> ()' to non-throwing function type '(Builtin.RawPointer) -> ()'
157 |       let buffer = Slab<count, Element>._initializationBuffer(start: rawPtr)
158 | 
```

@MaxDesiatov Should I create a bug towards this? Seems like we'd want this to compile properly, right? Or is there a way to disable the new `Slab.swift` stuff that's being added into the stdlib for now?

finagolfin commented:
> I did a LOT more testing and realized that using --host-target is not right for cross-compiling a Linux arch.

Yes, that is what I told you when you first tried it, that no other build preset uses that flag and while it was fine to [use it as a temporary fix, you should look into fixing the need for that internally instead](https://github.com/swiftlang/swift/pull/78991#issuecomment-2622604107).

> then the build scripts end up trying to use the built clang which fails since it was built for a different architecture

Makes sense, as that flag was only a temporary hack, to be used when cross-compiling the stdlib alone.

> it seems that using --cross-compile-hosts is correct for what I'm trying to do.

Yes and no, it is not the best to use initially when you're just cross-compiling the stdlib, but that flag is required later to cross-compile the corelibs, like Foundation.

> However, I am not opposed to the idea of kicking off cross-compilation in a different way, similar to android.

There's nothing different, [my Android CI uses that `--cross-compile-hosts` flag also](https://github.com/finagolfin/swift-android-sdk/blob/145f7671ee554b92873578ab71ea4f8e94c07035/.github/workflows/sdks.yml#L159).

> Perhaps by passing a linux-cross-compile or similar attribute which would enable cross-compiling specifically for Linux. Let me know your thoughts on this.

I don't see the need for a new flag: the current build-script has been cross-compiling to linux and Android for a long time with the existing flags.

> here's what I have for building just the Swift stdlib for armv7.

Seems fine, looks similar to my Android CI command linked above that I've been using for years now, except you have to disable those embedded/backtracing flags for linux also.

Looks like you and Max are still getting this pull working, let me know when it's ready and I will review. One thing you'll want to watch out for is that your cross-compilation changes don't break the existing linux toolchain build.

xtremekforever committed:
Find --gcc-install-dir in usr directory of sysroot

 - This is where it will be installed, and we don't want to conflict with other gcc installations at /home or otherwise.

xtremekforever commented:
> Looks like you and Max are still getting this pull working, let me know when it's ready and I will review. One thing you'll want to watch out for is that your cross-compilation changes don't break the existing linux toolchain build.

Thanks. That's a big deal, I want to ensure that all of these changes will not affect any other builds that we currently have. Thus, why it makes sense to introduce the `--cross-compile-sysroot` option since it is not used anywhere else.

We'll see what the solution can be to the other compilation issues. Worst case perhaps we can add a patch for those if cross-compilation them is an issue. Idk...

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

xtremekforever commented:
> @MaxDesiatov Should I create a bug towards this? Seems like we'd want this to compile properly, right? Or is there a way to disable the new `Slab.swift` stuff that's being added into the stdlib for now?

Looks like it builds now- I updated my nightly-jammy container to the latest snapshot and Slab.swift now builds. Phew. That answers that.

xtremekforever commented:
@MaxDesiatov, are you able to shed some light on this build error?

```
Error: Error: swift.exe exited with code 1.
Invocation:
  T:\Program Files\Swift\Toolchains\0.0.0+Asserts\usr\bin\swift.exe test --scratch-path T:\x64\sourcekit-lsp --package-path C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\sourcekit-lsp -c release -Xbuild-tools-swiftc -IT:\Program Files\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift -Xbuild-tools-swiftc -LT:\Program Files\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift\windows -Xcc -IT:\Program Files\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift -Xlinker -LT:\Program Files\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift\windows -debug-info-format none --parallel -Xcc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift-corelibs-libdispatch -Xcc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift-corelibs-libdispatch\src\BlocksRuntime -Xswiftc -IT:\5\lib\swift\host -Xswiftc -LT:\5\lib\swift\host -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\cmark\src\include -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\cmark\extensions\include -Xlinker -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\cmark\extensions\include -Xlinker T:\x64\cmark-gfm-0.29.0.gfm.13\src\cmark-gfm.lib -Xlinker T:\x64\cmark-gfm-0.29.0.gfm.13\extensions\cmark-gfm-extensions.lib -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift-system\Sources\CSystem\include -Xswiftc -IT:\20\swift -Xlinker -LT:\20\lib -Xswiftc -IT:\12\swift -Xlinker -LT:\12\lib -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\llbuild\products\libllbuild\include -Xswiftc -IT:\13\products\llbuildSwift -Xlinker -LT:\13\lib -Xswiftc -IT:\14\swift -Xlinker -LT:\14\lib -Xswiftc -IT:\16\swift -Xlinker -LT:\16\lib -Xlinker T:\16\lib\CCryptoBoringSSL.lib -Xswiftc -IT:\21\swift -Xlinker -LT:\21\lib -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift-markdown\Sources\CAtomic\inclde -Xlinker T:\22\lib\CAtomic.lib -Xswiftc -IT:\22\swift -Xlinker -LT:\22\lib -Xswiftc -IT:\23\swift -Xlinker -LT:\23\lib -Xswiftc -IT:\25\swift -Xlinker -LT:\25\Sources\IndexStoreDB -Xlinker T:\25\lib\CIndexStoreDB\CIndexStoreDB.lib -Xlinker T:\25\lib\Core\Core.lib -Xlinker T:\25\lib\Database\Database.lib -Xlinker T:\25\lib\Index\Index.lib -Xlinker T:\25\lib\LLVMSupport\LLVMSupport.lib -Xlinker T:\25\lib\Support\Support.lib -Xlinker T:\24\lib\CLMDB.lib -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\sourcekit-lsp\Sources\CAtomics\include -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\sourcekit-lsp\Sources\CSourcekitd\include -Xlinker T:\26\lib\CSourcekitd.lib -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\sourcekit-lsp\Sources\CCompletionScoring\include -Xswiftc -IT:\26\swift -Xlinker -LT:\26\lib
Call stack:
  at Invoke-Program, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 591
  at <ScriptBlock>, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 1411
  at Isolate-EnvVars, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 610
  at Build-SPMProject, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 1372
  at <ScriptBlock>, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 2798
  at Isolate-EnvVars, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 610
  at Test-SourceKitLSP, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 2785
  at <ScriptBlock>, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 3148

    at Invoke-Program, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 595
    at <ScriptBlock>, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 1411
    at Isolate-EnvVars, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 610
    at Build-SPMProject, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 1372
    at <ScriptBlock>, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 2798
    at Isolate-EnvVars, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 610
    at Test-SourceKitLSP, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 2785
    at <ScriptBlock>, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 3148
  From System.Management.Automation.RuntimeException: Error: swift.exe exited with code 1.
  Invocation:
    T:\Program Files\Swift\Toolchains\0.0.0+Asserts\usr\bin\swift.exe test --scratch-path T:\x64\sourcekit-lsp --package-path C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\sourcekit-lsp -c release -Xbuild-tools-swiftc -IT:\Program Files\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift -Xbuild-tools-swiftc -LT:\Program Files\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift\windows -Xcc -IT:\Program Files\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift -Xlinker -LT:\Program Files\Swift\Platforms\Windows.platform\Developer\SDKs\Windows.sdk\usr\lib\swift\windows -debug-info-format none --parallel -Xcc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift-corelibs-libdispatch -Xcc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift-corelibs-libdispatch\src\BlocksRuntime -Xswiftc -IT:\5\lib\swift\host -Xswiftc -LT:\5\lib\swift\host -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\cmark\src\include -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\cmark\extensions\include -Xlinker -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\cmark\extensions\include -Xlinker T:\x64\cmark-gfm-0.29.0.gfm.13\src\cmark-gfm.lib -Xlinker T:\x64\cmark-gfm-0.29.0.gfm.13\extensions\cmark-gfm-extensions.lib -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift-system\Sources\CSystem\include -Xswiftc -IT:\20\swift -Xlinker -LT:\20\lib -Xswiftc -IT:\12\swift -Xlinker -LT:\12\lib -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\llbuild\products\libllbuild\include -Xswiftc -IT:\13\products\llbuildSwift -Xlinker -LT:\13\lib -Xswiftc -IT:\14\swift -Xlinker -LT:\14\lib -Xswiftc -IT:\16\swift -Xlinker -LT:\16\lib -Xlinker T:\16\lib\CCryptoBoringSSL.lib -Xswiftc -IT:\21\swift -Xlinker -LT:\21\lib -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift-markdown\Sources\CAtomic\inclde -Xlinker T:\22\lib\CAtomic.lib -Xswiftc -IT:\22\swift -Xlinker -LT:\22\lib -Xswiftc -IT:\23\swift -Xlinker -LT:\23\lib -Xswiftc -IT:\25\swift -Xlinker -LT:\25\Sources\IndexStoreDB -Xlinker T:\25\lib\CIndexStoreDB\CIndexStoreDB.lib -Xlinker T:\25\lib\Core\Core.lib -Xlinker T:\25\lib\Database\Database.lib -Xlinker T:\25\lib\Index\Index.lib -Xlinker T:\25\lib\LLVMSupport\LLVMSupport.lib -Xlinker T:\25\lib\Support\Support.lib -Xlinker T:\24\lib\CLMDB.lib -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\sourcekit-lsp\Sources\CAtomics\include -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\sourcekit-lsp\Sources\CSourcekitd\include -Xlinker T:\26\lib\CSourcekitd.lib -Xswiftc -IC:\Users\swift-ci\jenkins\workspace\swift-PR-windows\sourcekit-lsp\Sources\CCompletionScoring\include -Xswiftc -IT:\26\swift -Xlinker -LT:\26\lib
  Call stack:
    at Invoke-Program, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 591
    at <ScriptBlock>, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 1411
    at Isolate-EnvVars, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 610
    at Build-SPMProject, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 1372
    at <ScriptBlock>, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 2798
    at Isolate-EnvVars, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 610
    at Test-SourceKitLSP, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 2785
    at <ScriptBlock>, C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\swift\utils\build.ps1: line 3148
```

I'm not sure how to approach fixing this if it's caused by this PR.

MaxDesiatov commented:
@swift-ci test windows

xtremekforever commented:
@finagolfin @MaxDesiatov looks like it built! 🎉

Let me know if these changes will be acceptable. My next work will be to ensure the core libs- dispatch, foundation, xctest, and swift testing build with a Linux cross compilation profile. We already know that Swift 6 and later builds for armv7, so now it comes down to supporting it from the build scripts 😁

finagolfin commented on utils/build-script-impl:
```diff
@@ -288,6 +288,7 @@ KNOWN_SETTINGS=(
     cross-compile-with-host-tools                 ""                "set to use the clang we build for the host to then build the cross-compile hosts"
     cross-compile-install-prefixes                ""                "semicolon-separated list of install prefixes to use for the cross-compiled hosts. The list expands, so if there are more cross-compile hosts than prefixes, unmatched hosts use the last prefix in the list"
     cross-compile-deps-path                       ""                "path for CMake to look for cross-compiled library dependencies, such as libXML2"
+    cross-compile-sysroot                         ""                "path for CMake to look for for a cross-compilation sysroot for a different platform and/or architecture"
```
"for for"

finagolfin commented on utils/build-script-impl:
```diff
@@ -1109,7 +1110,7 @@ function false_true() {
 for t in "${CROSS_COMPILE_HOSTS[@]}"; do
     case ${t} in
-        macosx* | iphone* | appletv* | watch* | linux-armv5 | linux-armv6 | linux-armv7 | android-* | openbsd-* | linux-static-* )
+        macosx* | iphone* | appletv* | watch* | linux-* | android-* | openbsd-* | linux-static-* )
```
This makes the later "linux-static-*" redundant, maybe just change yours to "linux-arm*" for now?

xtremekforever replied:
I'm opening up cross-compiling to any linux- arch here! I see great benefit in being able to cross-compile from one arch to another easily, even if that arch is officially supported by Swift ;) I will remove the redundant case.

finagolfin commented on utils/build-script-impl:
```diff
@@ -1743,6 +1744,7 @@ for host in "${ALL_HOSTS[@]}"; do
                     )
                 fi
 
+
```
Remove

finagolfin commented on utils/build-script-impl:
```diff
@@ -1779,6 +1781,16 @@ for host in "${ALL_HOSTS[@]}"; do
+                        cmake_options=(
+                            "${cmake_options[@]}"
+                            -DCROSS_COMPILE_SYSROOT:STRING="${CROSS_COMPILE_SYSROOT}"
+                            -DSWIFT_USE_LINKER:STRING="lld"
```
Why force this linker here?

xtremekforever replied:
I found that `gold` struggles to link libraries when using a different arch from the host. It'll just fail when CMake attempts to see that clang can compile programs for the given target with an error about not supporting the platform. If you have a solution where we could get `gold` to work I'm all ears.

xtremekforever replied:
Here's an example. I removed the lld line and got:

```
CMake Error at /usr/share/cmake-3.28/Modules/CMakeTestCCompiler.cmake:67 (message):
  The C compiler

    "~/.local/share/swiftly/bin/clang"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: '~/swift-main-build/build/build-armv7/swift-linux-armv7/libdispatch-linux-armv7-prefix/src/libdispatch-linux-armv7-build/CMakeFiles/CMakeScratch/TryCompile-XosJG0'
    
    Run Build Command(s): /usr/bin/ninja -v cmTC_a0ad5
    [1/2][ 50%][0.071s] ~/.local/share/swiftly/bin/clang --target=armv7-unknown-linux-gnueabihf --sysroot=~/swift-main-build/sysroot-debian-bookworm   -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fno-semantic-interposition -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -w -fuse-ld=gold -MD -MT CMakeFiles/cmTC_a0ad5.dir/testCCompiler.c.o -MF CMakeFiles/cmTC_a0ad5.dir/testCCompiler.c.o.d -o CMakeFiles/cmTC_a0ad5.dir/testCCompiler.c.o -c ~/swift-main-build/build/build-armv7/swift-linux-armv7/libdispatch-linux-armv7-prefix/src/libdispatch-linux-armv7-build/CMakeFiles/CMakeScratch/TryCompile-XosJG0/testCCompiler.c
    [2/2][100%][0.125s] : && ~/.local/share/swiftly/bin/clang --target=armv7-unknown-linux-gnueabihf --sysroot=~/swift-main-build/sysroot-debian-bookworm -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fno-semantic-interposition -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -w -fuse-ld=gold  CMakeFiles/cmTC_a0ad5.dir/testCCompiler.c.o -o cmTC_a0ad5   && :
    FAILED: cmTC_a0ad5 
    : && ~/.local/share/swiftly/bin/clang --target=armv7-unknown-linux-gnueabihf --sysroot=~/swift-main-build/sysroot-debian-bookworm -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -fPIC -fno-semantic-interposition -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -w -fuse-ld=gold  CMakeFiles/cmTC_a0ad5.dir/testCCompiler.c.o -o cmTC_a0ad5   && :
    /usr/bin/ld.gold: fatal error: ~/swift-main-build/sysroot-debian-bookworm/lib/arm-linux-gnueabihf/Scrt1.o: unsupported ELF machine number 40
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    ninja: build stopped: subcommand failed.
```

MaxDesiatov replied:
Yeah, `gold` is supposed to distinguish target triples via command name prefixes, so plain `ld.gold` invocation is not supposed to cross-compile IIUC, you'd have to use something like `armv7-linux-ld` or whatever.

finagolfin commented on utils/build_swift/build_swift/driver_arguments.py:
```diff
@@ -669,6 +669,12 @@ def create_argument_parser():
                 'such as the libcurl dependency of FoundationNetworking')
 
+    option('--cross-compile-sysroot', store_path,
+           help='The path to a directory that contains a sysroot for a different '
```
sysroot -> C/C++ sysroot

finagolfin commented on utils/build_swift/build_swift/driver_arguments.py:
```diff
@@ -669,6 +669,12 @@ def create_argument_parser():
 
+    option('--cross-compile-sysroot', store_path,
+           help='The path to a directory that contains a sysroot for a different '
+                'platform and/or architecture for cross-compiling Swift to. '
```
Swift to -> Swift

finagolfin commented on utils/build_swift/build_swift/driver_arguments.py:
```diff
@@ -669,6 +669,12 @@ def create_argument_parser():
+    option('--cross-compile-sysroot', store_path,
+           help='The path to a directory that contains a sysroot for a different '
+                'platform and/or architecture for cross-compiling Swift to. '
+                'The sysroot must contain libraries that match the architecture '
```
libraries -> C/C++ libraries

Helps to be more specific 😺

finagolfin commented on utils/swift_build_support/swift_build_support/products/product.py:
```diff
@@ -479,6 +481,11 @@ def common_cross_c_flags(self, platform, arch, include_arch=False):
+        # Use lld if external sysroot is provided
+        if (self.is_cross_compile_target('{}-{}'.format(platform, arch))
+                and self.args.cross_compile_sysroot):
+            cross_flags.append('-w -fuse-ld=lld')
```
Maybe just set this for linux-armv7 for now?

xtremekforever replied:
I see that even if I'm cross-compiling from a linux-x86_64 host to linux-aarch64, `gold` still fails to find dependencies in the external sysroot but `lld` works. So, that is why I am defaulting to using `lld` if the `--cross-compile-sysroot` is passed, and you're cross compiling.

xtremekforever replied:
Here's an example with a debian-bookworm sysroot for arm64 if I disable the use of lld:

```
CMake Error at /usr/share/cmake-3.28/Modules/CMakeTestCCompiler.cmake:67 (message):
  The C compiler

    "~/.local/share/swiftly/bin/clang"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: '~/swift-main-build/build/build-aarch64/llvm-linux-aarch64/CMakeFiles/CMakeScratch/TryCompile-QU4odf'
    
    Run Build Command(s): /usr/bin/ninja -v cmTC_558a0
    [1/2][ 50%][0.027s] ~/.local/share/swiftly/bin/clang --target=aarch64-unknown-linux-gnu --sysroot=~/swift-main-build/sysroot-debian-bookworm-arm64   -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector -MD -MT CMakeFiles/cmTC_558a0.dir/testCCompiler.c.o -MF CMakeFiles/cmTC_558a0.dir/testCCompiler.c.o.d -o CMakeFiles/cmTC_558a0.dir/testCCompiler.c.o -c ~/swift-main-build/build/build-aarch64/llvm-linux-aarch64/CMakeFiles/CMakeScratch/TryCompile-QU4odf/testCCompiler.c
    [2/2][100%][0.050s] : && ~/.local/share/swiftly/bin/clang --target=aarch64-unknown-linux-gnu --sysroot=~/swift-main-build/sysroot-debian-bookworm-arm64 -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector  CMakeFiles/cmTC_558a0.dir/testCCompiler.c.o -o cmTC_558a0   && :
    FAILED: cmTC_558a0 
    : && ~/.local/share/swiftly/bin/clang --target=aarch64-unknown-linux-gnu --sysroot=~/swift-main-build/sysroot-debian-bookworm-arm64 -Wno-unknown-warning-option -Werror=unguarded-availability-new -fno-stack-protector  CMakeFiles/cmTC_558a0.dir/testCCompiler.c.o -o cmTC_558a0   && :
    /usr/bin/ld.gold: fatal error: ~/swift-main-build/sysroot-debian-bookworm-arm64/lib/aarch64-linux-gnu/Scrt1.o: unsupported ELF machine number 183
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    ninja: build stopped: subcommand failed.
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:45 (project)
```

Hopefully this gives you context.

xtremekforever committed:
Rename cross-compile-sysroot -> cross-compile-sysroots and update wording

xtremekforever committed:
Remove redundant case for linux-static-*, remove extra whitespace

xtremekforever commented on cmake/modules/SwiftConfigureSDK.cmake:
```diff
@@ -407,14 +407,33 @@ macro(configure_sdk_unix name architectures)
+          if(FOUND_GCC_INSTALL_DIR)
+            set(SWIFT_SDK_${prefix}_CXX_OVERLAY_SWIFT_COMPILE_FLAGS -Xcc --gcc-install-dir=${GCC_INSTALL_DIR})
+          else()
+            set(SWIFT_SDK_${prefix}_CXX_OVERLAY_SWIFT_COMPILE_FLAGS -Xcc --gcc-toolchain=${CROSS_COMPILE_SYSROOTS})
```
Yes so `--gcc-toolchain` is used to instruct the compiler where to find libstdc++ for the CxxStdlib module. Without this, we have no CXX integration. If you look further up in this cmake file there's a whole block talking about how the `--gcc-toolchain` is hardcoded to `/usr` because of a bug. However, when we are using an external sysroot, that path needs to change to point to the sysroot instead so CxxStdlib can be built.

I can add some comments into the code explaining this 👍🏻

MaxDesiatov replied:
yes, more contextual in-place comments please, IMO it's not obvious enough one would need to look further up in this file

xtremekforever replied:
Actually...I just ran some tests clearing these flags with an external sysroot, and it seems like the compiler is not having a problem finding the GCC installation to build CxxStdlib!! If this works consistently with different OS versions, it may make more sense to _clear_ this flag when using an external sysroot, but then if needed for a specific sysroot someone can pass the `--gcc-toolchain` or `--gcc-install-dir` as an override for non-standard installation directories. I'll go ahead and do this since it is much simpler.

xtremekforever committed:
Remove --gcc-install-dir and --gcc-toolchain code when using external sysroot

 - Turns out the compiler and linker are capable of finding libstdc++ with all versions of Ubuntu/Debian sysroots passed to it, so we use that instead of the hardcoded --gcc-toolchain for regular builds.

xtremekforever committed:
Restore original code for setting SWIFT_SDK_LINUX_ARCH_${arch}_TRIPLE for LINUX

MaxDesiatov commented:
CC @kateinoigakukun if there's ever an interest to cross-compile the toolchain itself with WASI-libc.

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

kateinoigakukun commented on cmake/modules/Libdispatch.cmake:
```diff
@@ -75,6 +75,12 @@ foreach(sdk ${DISPATCH_SDKS})
+      set(SWIFT_LIBDISPATCH_COMPILER_CMAKE_ARGS ${SWIFT_LIBDISPATCH_COMPILER_CMAKE_ARGS};-DCMAKE_SYSROOT=${SWIFT_SDK_${sdk}_ARCH_${arch}_PATH})
+      set(SWIFT_LIBDISPATCH_COMPILER_TRIPLE_CMAKE_ARGS -DCMAKE_C_COMPILER_TARGET=${SWIFT_SDK_${sdk}_ARCH_${arch}_TRIPLE};-DCMAKE_CXX_COMPILER_TARGET=${SWIFT_SDK_${sdk}_ARCH_${arch}_TRIPLE})
+      # Use linker used to build Swift
+      set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -w -fuse-ld=${SWIFT_USE_LINKER}")
```
This `Libdispatch.cmake` is loaded by `include(Libdispatch)`, so we should not overwrite `CMAKE_C_FLAGS` here. Could you define a new local scope variable instead?

kateinoigakukun commented on utils/build-script-impl:
```diff
@@ -1779,6 +1780,16 @@ for host in "${ALL_HOSTS[@]}"; do
+                    if [ ! -z "${CROSS_COMPILE_SYSROOTS}" ]; then
+                        cmake_options=(
+                            "${cmake_options[@]}"
+                            -DCROSS_COMPILE_SYSROOTS:STRING="${CROSS_COMPILE_SYSROOTS}"
```
Could you add `SWIFT_` prefix as well as others to indicate the variable is defined by Swift project?

xtremekforever replied:
So to be clear, just add the SWIFT_ prefix to the variable used to pass to Swift cmake invocation? SWIFT_CROSS_COMPILE_SYSROOTS

I will note that the CROSS_COMPILE_SYSROOTS will be used down the line to also tell the core libs how to cross compile, and possibly other things as well. So I would leave the local param as --cross-compile-sysroots.

kateinoigakukun replied:
Yes, it's fine to keep the option name of build-script without having prefix

xtremekforever committed:
Add SWIFT_ prefix to CROSS_COMPILE_SYSROOTS when passed to Swift cmake invocation

xtremekforever committed:
Implement new SWIFT_LIBDISPATCH_C_FLAGS and SWIFT_LIBDISPATCH_CXX_FLAGS variables in Libdispatch.cmake

 - So we do not overwrite CMAKE_C_FLAGS and CMAKE_CXX_FLAGS

MaxDesiatov commented:
@swift-ci test

MaxDesiatov commented:
@swift-ci build toolchain

finagolfin commented on utils/build-script-impl:
```diff
@@ -288,6 +288,7 @@ KNOWN_SETTINGS=(
     cross-compile-with-host-tools                 ""                "set to use the clang we build for the host to then build the cross-compile hosts"
     cross-compile-install-prefixes                ""                "semicolon-separated list of install prefixes to use for the cross-compiled hosts. The list expands, so if there are more cross-compile hosts than prefixes, unmatched hosts use the last prefix in the list"
     cross-compile-deps-path                       ""                "path for CMake to look for cross-compiled library dependencies, such as libXML2"
+    cross-compile-sysroots                        ""                "path for CMake to look for a cross-compilation sysroot for a different platform and/or architecture"
```
"for a cross-compilation sysroot for a different platform and/or architecture" -> for cross-compilation sysroots for different platforms or architectures

xtremekforever replied:
Okay, I'll wait to see if anyone else has a comment and then implement

MaxDesiatov replied:
yes, plural/singular inconsistency reads weird to me here

xtremekforever committed:
Update wording for cross-compile-sysroots option

xtremekforever commented:
@MaxDesiatov wording changes done. Let me know if there will be anything else! I have tested this PR with building against a variety of Ubuntu/Debian sysroots and arches and it seems to work really well, so now it comes down to any other reviewers who want to comment I guess...

I'm now looking at building the corelibs and trying to figure out the best way to update the scripts to pass the correct flags, etc. @finagolfin can I get in touch with you again to ask you a few questions...? This is related to the targets.py stuff that we had talked about before related to `AndroidPlatform`, and to bounce some ideas off of you if that's okay.

finagolfin commented:
Sure, the best way to reach me is my email, which you can get from my github patches. If you plan to use the CMake toolchain file, I don't know much about that format and the current Android build does not use it, but the flags both set should be pretty similar.

MaxDesiatov commented:
@swift-ci build toolchain

MaxDesiatov commented:
@swift-ci test

xtremekforever commented:
@MaxDesiatov another fluke on the macOS run? I saw it last time where watchOS failed to build, but then it worked the next time you ran the CI.

finagolfin commented:
@swift-ci please test macos

xtremekforever commented:
@etcwilde @al45tair any input into this? It's passed the CI multiple times and seems to be pretty solid thus far. I would like to see this included soon so I can continue to the next step to supporting Linux cross-compilation in the build scripts.

xtremekforever commented:
@edymtt @MaxDesiatov This has been sitting for a few days...I know this doesn't seem like a very exciting PR to review and that you guys are busy, but I'd like to see a little progress here. Maybe it'll help to show off what these small changes can do just inside of a container, and what is generated with the following commands:

```bash
$ docker run -it --name swift-armv7-builder -v $(pwd):/src swiftlang/swift:nightly-jammy /bin/bash
$ apt update && apt install cmake ninja-build libsqlite3-dev ncurses-dev git python3 libstdc++-11-dev-armhf-cross
$ cd /src && export SWIFT_NATIVE_PATH=/usr/bin
$ ./swift/utils/build-script --cross-compile-hosts=linux-armv7 --cross-compile-sysroot=/ \
     --build-llvm=0  --skip-local-build --build-swift-tools=0 --swift-enable-backtracing=0 --build-embedded-stdlib=0 \
     --native-swift-tools-path=$SWIFT_NATIVE_PATH --native-clang-tools-path=$SWIFT_NATIVE_PATH \
     --extra-cmake-options="-DSWIFT_ENABLE_RUNTIME_MODULE=FALSE" --release --install-swift --install-destdir=swift-armv7-install
```

I ran the commands above on my MacBook Pro / M1 Pro. The following artifacts are generated:

```bash
linux-armv7/usr/lib/swift/linux
├── Cxx.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── CxxStdlib.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── Distributed.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── Glibc.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── Observation.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── RegexBuilder.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── Swift.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── SwiftOnoneSupport.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── Synchronization.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── _Builtin_float.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── _Concurrency.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── _Differentiation.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── _RegexParser.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── _StringProcessing.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── _Volatile.swiftmodule
│   ├── armv7-unknown-linux-gnueabihf.abi.json
│   ├── armv7-unknown-linux-gnueabihf.private.swiftinterface
│   ├── armv7-unknown-linux-gnueabihf.swiftdoc
│   ├── armv7-unknown-linux-gnueabihf.swiftinterface
│   └── armv7-unknown-linux-gnueabihf.swiftmodule
├── armv7
│   ├── SwiftGlibc.h
│   ├── glibc.modulemap
│   └── swiftrt.o
├── libcxxshim.h
├── libcxxshim.modulemap
├── libcxxstdlibshim.h
├── libstdcxx.h
├── libstdcxx.modulemap
├── libswiftCore.so
├── libswiftCxx.a
├── libswiftCxxStdlib.a
├── libswiftDistributed.so
├── libswiftGlibc.so
├── libswiftObservation.so
├── libswiftRegexBuilder.so
├── libswiftRemoteMirror.so
├── libswiftSwiftOnoneSupport.so
├── libswiftSynchronization.so
├── libswift_Builtin_float.so
├── libswift_Concurrency.so
├── libswift_Differentiation.so
├── libswift_RegexParser.so
├── libswift_StringProcessing.so
└── libswift_Volatile.so
```

Everything has the correct arch, and I can run it on the target (in my case a Raspberry Pi 2B) if I also manually scp the `libdispatch.so` and `libBlocksRuntime.so` files as well. It's not very useful without also building libdispatch and Foundation, but it works!

Can we get this completed so that this initial support is in, then I can work on the next step?

finagolfin commented:
@xtremekforever, the majority of this pull simply adding the new `--cross-compile-sysroot` flag should be fine, but I just wanted to get some of the other changes, like making lld the default in these scenarios, reviewed by others who actually build for linux with the CMake toolchain file, which I don't.

I see no reason you cannot continue on to building your next steps locally in the meantime: are you worried this will be rejected completely or what? [The policy here is weekly pings](https://github.com/swiftlang/swift/blob/main/docs/HowToGuides/FirstPullRequest.md#i-didnt-get-a-response-from-someone-what-should-i-do), but I advise you that given how busy everybody is and how niche 32-bit armv7 is nowadays, this pull may take a little longer.

One step you may want to take next that would give us more info and may help validate this pull is to run the host tests from the compiler validation suite with your new armv7 cross-compilation config. Take a look at [the flags for how this is done for Android in the doc,](https://github.com/swiftlang/swift/blob/main/docs/Android.md), add the `--host-test` flag mentioned there, and try that out for linux armv7.

Basically, you'd build the Swift compiler and some other testing tools also by removing the `--cross-compile-hosts=linux-armv7 --build-llvm=0 --skip-local-build --build-swift-tools=0 --native-swift-tools-path=$SWIFT_NATIVE_PATH --native-clang-tools-path=$SWIFT_NATIVE_PATH` flags from your Docker command and add something like `-T --stdlib-deployment-targets=linux-armv7` instead. If you're still seeing the `host-target` issue you mentioned in your last pull for this too, we can work through that over email.

xtremekforever commented:
@finagolfin thanks for the reply! My only worry is that it gets forgotten and by the time it's revisited we're already release Swift 6.2, etc. So I want to keep pinging on it, and if there's anything else to do to improve this or make it easier to approve I'll do it! I intend on seeing this through until the end.

Let me try your suggestion to run some tests! I'm not sure if it'll work, but I'll let you know. I did send you an email the other day but got no response- is finagolfin@tuta.io the correct email?

finagolfin commented:
Yes, I got your email, just been a bit busy this week, will get back to you soon.

xtremekforever commented:
@finagolfin okay no problem! I just wasn't sure if you had gotten it. You don't have to worry about that one since it was just me asking what you thought.

I tried running this command:

```
./swift/utils/build-script --stdlib-deployment-targets=linux-armv7 -T --swift-enable-backtracing=0 --build-embedded-stdlib=0 --extra-cmake-options="-DSWIFT_ENABLE_RUNTIME_MODULE=FALSE"
```

And as expected, it doesn't work:

```
+ /usr/bin/cmake --build ~/swift-main-build/build/Ninja-DebugAssert/swift-linux-x86_64 -- -j12 all swift-test-stdlib-linux-armv7 swift-libexec-linux-armv7
ninja: error: unknown target 'swift-test-stdlib-linux-armv7'
```

The scripts don't know how to compile for linux-armv7...as happened last time. Any ideas? Maybe we can continue this over email.

xtremekforever committed:
Merge branch 'main' into #78960-swift-stdlib-cross-compile

edymtt commented:
Apologies for my late reply here -- I am currently not able to find the time to review this properly for two main reasons:
* as @finagolfin correctly assumed, I have been involved in multiple projects;
* I am not very familiar with how we build under Linux (I have more experience building for Darwin platforms) -- so in a sense for me this PR is not straightforward to review, as I have to learn the underlying concepts, catch up with the existing architecture and figure out how the PR fits in that.

I will try to chip at this, but I can make no promises on when I will able to complete my review (assuming my priorities stay the same)

xtremekforever commented:
> Apologies for my late reply here -- I am currently not able to find the time to review this properly for two main reasons:
> 
> * as @finagolfin correctly assumed, I have been involved in multiple projects;
> * I am not very familiar with how we build under Linux (I have more experience building for Darwin platforms) -- so in a sense for me this PR is not straightforward to review, as I have to learn the underlying concepts, catch up with the existing architecture and figure out how the PR fits in that.
> 
> I will try to chip at this, but I can make no promises on when I will able to complete my review (assuming my priorities stay the same)

Thanks for the reply. I understand that you're busy. I'll ping here again in a few weeks if nothing happens to see what's going on. In the meantime tho, while I wait for more review, I'll be working on other details like getting the `backtracing` module to compile properly for armv7, and setting up cross-compilation of the corelibs using CMake toolchain files (this will be a separate PR). My only hope is we can get this merged in time for 6.2, so that the Swift Everywhere ecosystem can keep expanding! 👍🏻

edymtt commented:
Just a quick update that I am still busy with other efforts, so I did not get a chance to ponder on this PR.

xtremekforever commented:
> Just a quick update that I am still busy with other efforts, so I did not get a chance to ponder on this PR.

Thanks for the ping. I've been thinking about these changes myself and wondering what kind of work you and @etcwilde have been making to the build system- is there any info you guys can point me to like a forum post or a branch so I can see what's in progress? I'd like to understand where things are moving, and maybe that could influence these changes as well. If you have a moment...

finagolfin commented:
I believe you can see [all the new work they're doing to change how the stdlib is built as it's merged here](https://github.com/swiftlang/swift/tree/main/Runtimes).

etcwilde commented:
So frankly, I'm not sure how to get the external sysroot to work without stepping on someone's toes (or some configuration that someone is using anyway). If we had separate scripts or at least some separation between building the compiler and the runtimes, it would be easier since all of the products in each phase should build against the same sysroot as they are ending up in the same place. As you have probably noticed, some configurations are using the just-built compiler to build the runtimes, but if the runtimes exist in the build directory, use the just-built runtimes instead. If you've cross-compiled the runtimes, that doesn't work. I'm working on a vision update to share with the community on where I'm hoping to see things end up.

So in terms of direction of the new build. It's about breaking apart the projects based on where they need to land and what runtimes they need available to build. The first step is splitting the runtime/stdlib build out of the compiler build so that they are separate CMake projects entirely. I've added facilities to make it easier to build it with build-script, but that is for local development convenience more than for shipping. The intent is that you can point CMake at `Runtimes/Core` and have it build you a standard library. There are knobs to enable and disable features depending on your needs, or the needs of the vendor, rather than hard-coding things based on a specific configuration. Then we use CMake cache files for re-producing specific build configurations. e.g. if you wanted to build the Swift runtime how Apple builds it for the arm64 iOS simulator, you would point CMake at the [arm64-iPhoneOS-simulator cache](https://github.com/swiftlang/swift/blob/main/Runtimes/Core/cmake/caches/Vendors/Apple/arm64-iPhoneOS-simulator.cmake).

By using pure CMake, we're also able to take advantage of existing infrastructure and processes. This invocation should cross-compile the libswiftCore as a static archive for Android, for instance.

```sh
cmake -B build-android -S swift/Runtimes/Core -G 'Ninja' \ 
 -DCMAKE_Swift_COMPILER_WORKS=YES \
 -DCMAKE_Swift_COMPILER=/Users/ewilde/Work/Swift-Project/build/Ninja-ReleaseAssert/swift-macosx-arm64/bin/swiftc \
 -DCMAKE_Swift_COMPILER_TARGET=aarch64-linux-android31 \
 --toolchain /Users/ewilde/AndroidNDK11579264.app/Contents/NDK/build/cmake/android.toolchain.cmake \
 -DANDROID_ABI=arm64-v8a \
 -DANDROID_PLATFORM=android-31
cmake --build build-android
```

The toolchain and ABI flags come from the [Android documentation](https://developer.android.com/ndk/guides/cmake#variables) from Google so there's no guessing about what they do.
The Android NDK doesn't know about Swift, and the Swift runtimes require a matching Swift compiler, so there are a few flags that need to be set to get the Swift bits to build, but we can explain exactly why each one of them exists.
 - `CMAKE_Swift_COMPILER_WORKS`: The CMake compiler tests require a standard library on the platform you're building for. The NDK doesn't have a Swift standard library, and we're in the process of building it, so those tests will fail because there isn't a standard library.
 - `CMAKE_Swift_COMPILER`: Pointed at my just-built compiler. If I have one that is new enough in my toolchain, I can remove this. The Android NDK toolchain file sets the C and C++ compilers automatically, which is why there aren't override for that.
 - `CMAKE_Swift_COMPILER_TARGET`: The Android NDK doesn't know about Swift so it doesn't set the target triple to pass to Swift. This sets it.

From there, we can choose to enable or disable pieces. `ccmake build-android` is your friend for enabling and disabling bits.

The readme in `Runtimes/` has some info on getting started with the new build and how things are organized and the readme in `Runtimes/Core` has info on how to build libswiftCore and the other Core libraries.

Hope that helps.

xtremekforever commented:
So I'm taking a step back here for a bit to see if there's a better way to do this without stepping on anyone's workflows. One thing I've realized that would work decently well would be to use a CMake toolchain file to cross compile everything from llvm to Swift and the corelibs. This toolchain file would contain stuff like this for armv7, for example:

```cmake
set(CMAKE_CXX_COMPILER_TARGET armv7-unknown-linux-gnueabihf)
set(CMAKE_C_COMPILER_TARGET armv7-unknown-linux-gnueabihf)
set(CMAKE_SYSROOT /src/sysroot-debian-bookworm)
set(CMAKE_SYSTEM_NAME Linux)
set(CMAKE_SYSTEM_PROCESSOR armv7-a)
set(CMAKE_Swift_COMPILER_TARGET armv7-unknown-linux-gnueabihf)
```

The CMAKE_SYSTEM_PROCESSOR in particular is already handled by SwiftSupport.cmake in all the various corelibs:

```cmake
function(get_swift_host_arch result_var_name)
  if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86_64")
    set("${result_var_name}" "x86_64" PARENT_SCOPE)
  elseif ("${CMAKE_SYSTEM_PROCESSOR}" MATCHES "AArch64|aarch64|arm64|ARM64")
    if(CMAKE_SYSTEM_NAME MATCHES Darwin)
      set("${result_var_name}" "arm64" PARENT_SCOPE)
    else()
      set("${result_var_name}" "aarch64" PARENT_SCOPE)
    endif()
  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "ppc64")
    set("${result_var_name}" "powerpc64" PARENT_SCOPE)
  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "ppc64le")
    set("${result_var_name}" "powerpc64le" PARENT_SCOPE)
  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "s390x")
    set("${result_var_name}" "s390x" PARENT_SCOPE)
  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv6l")
    set("${result_var_name}" "armv6" PARENT_SCOPE)
  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv7-a")
    set("${result_var_name}" "armv7" PARENT_SCOPE)
  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "armv7l")
    set("${result_var_name}" "armv7" PARENT_SCOPE)
  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "AMD64")
    set("${result_var_name}" "x86_64" PARENT_SCOPE)
  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "IA64")
    set("${result_var_name}" "itanium" PARENT_SCOPE)
  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86")
    set("${result_var_name}" "i686" PARENT_SCOPE)
  elseif("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "i686")
    set("${result_var_name}" "i686" PARENT_SCOPE)
  else()
    message(FATAL_ERROR "Unrecognized architecture on host system: ${CMAKE_SYSTEM_PROCESSOR}")
  endif()
endfunction()
```

So you just give it the right value and it'll compile for your arch. I'm not sure if this would work for llvm and the Swift stdlib, but perhaps a better way to approach things instead of changing what's there. Any thoughts on this solution instead of changing the existing CMake files, @finagolfin and @edymtt ?

finagolfin commented:
[Both the linux and mac toolchain builds already use a CMake toolchain file](https://github.com/swiftlang/swift/blob/7205e447/utils/swift_build_support/swift_build_support/products/product.py#L438), both for native and cross-compilation of a handful of build products, including [the massive LLVM build](https://github.com/swiftlang/swift/blob/f68e23aaf1f35c6a9fdb544b7810384b08c598d6/utils/swift_build_support/swift_build_support/products/llvm.py#L245) and [the new libTesting](https://github.com/swiftlang/swift/blob/d092105e6622c3b46661193491f0c937eb406eda/utils/swift_build_support/swift_build_support/products/swift_testing.py#L118). I thought you were already using that linux toolchain file for this cross-compile, [hence my referencing it earlier in this very thread](https://github.com/swiftlang/swift/pull/79090#issuecomment-2660324831), but searching for it now, I don't see it currently being used to build the stdlib or corelibs.

If you want to use that file more, that is the direction others like Evan have expressed they'd like to move the build towards, so I suspect any pulls expanding and using that toolchain file more would be gladly accepted. I suggest you examine the Python code that currently generates the CMake toolchain files used more closely and compare it to the generated file, so you know exactly how the generation works, then modify the generation for your platform and use it more broadly, as you saw OpenBSD does.

There is no substitute for reading the code and asking questions: I have not received any emails from you for the last month, including after a request to see your failing CMake config output. I was recently told about a `--linux-archs` flag on the forum that the static linux SDK uses in its build: that may fix the build issue you saw when trying to run the compiler tests.

xtremekforever commented:
> Both the linux and mac toolchain builds already use a CMake toolchain file, both for native and cross-compilation of a handful of build products, including the massive LLVM build and the new libTesting. I thought you were already using that linux toolchain file for this cross-compile, hence my referencing it earlier in this very thread, but searching for it now, I don't see it currently being used to build the stdlib or corelibs.

Yeah I noticed that llvm does generate a cmake toolchain file for cross compilation, not sure if stdlib also does this, but for sure the corelibs do not- they are just built with cmake commands directly from the build-script-impl. So, to get them to use the swift build support products they'd all need to be brought in and built through that, which could be prone to a lot of errors...

finagolfin commented:
> they are just built with cmake commands directly from the build-script-impl. So, to get them to use the swift build support products they'd all need to be brought in and built through that, which could be prone to a lot of errors...

Would they? You've already noted that [OpenBSD just passes the toolchain file as an additional flag to this compiler/stdlib repo _in_ build-script-impl](https://github.com/swiftlang/swift/blob/d260bd8ddb4d6b6f91726ca7a2ac94051e531f27/utils/build-script-impl#L1652): why not try the same with linux armv7 for the stdlib and corelibs _in_ build-script-impl? Considering OpenBSD is already doing it for this repo, I doubt you'll have much of a problem with all four stdlib/corelibs repos, plus you know libTesting already builds using it.
